daml/ContingentClaims/Util.daml

```diff
@@ -23,20 +25,34 @@ import ContingentClaims.Claim (Claim, Claim(..), ClaimF(..))
 import ContingentClaims.Observation (Observation)
 import ContingentClaims.Observation qualified as O
 import DA.Foldable (Foldable, sum, fold)
-import Daml.Control.Arrow ((&&&), (>>>), (<<<), Kleisli(..))
+import Daml.Control.Arrow ((&&&), (|||), (>>>), (<<<), Kleisli(..))
 import Daml.Control.Recursion
-import Prelude hiding (sum, sequence)
-import DA.Traversable (Traversable, sequence)
+import Prelude hiding (sum, sequence, mapA)
+import DA.Traversable (Traversable, sequence, mapA)
 
 -- The morphisms ending in 'M' are monadic variants, allowing to interleave e.g. `Update` or `Script`.
+-- `cataM` After Tim Williams' talk, https://www.youtube.com/watch?v=Zw9KeP3OzpU.
+
+-- TODO: the two folds can probably be simplified with Traverse.mapA too.
 cataM : (Monad m, Traversable f, Recursive b f) => (f a -> m a) -> b -> m a
 cataM f b = (project >>> fmap (cataM f) >>> (>>= f) . sequence) b
 
 paraM : (Monad m, Traversable f, Recursive b f) => (f (b, a) -> m a) -> b -> m a
 paraM f b = (project >>> fmap (runKleisli $ Kleisli pure &&& Kleisli (paraM f)) >>> (>>= f) . sequence) b
 
+-- FIXME : I think the algebra should be a -> m (f a)
 anaM : (Monad m, Traversable f, Corecursive b f) => (a -> f (m a)) -> a -> m b
-anaM f a = (fmap embed <<< sequence <<< fmap (>>= anaM f) <<< f) a
+anaM f a = (fmap embed <<< mapA (>>= anaM f) <<< f) a
+
+apoM : (Monad m, Traversable f, Corecursive b f) => (a -> m (f (Either b a))) -> a -> m b
+apoM f a = (fmap embed <<< (>>= (mapA (pure ||| apoM f))) <<< f) a
+
+-- | Specialised lazy re-fold, used by `lifecycle`.
+apoCataM : (Monad m, Traversable f, Corecursive b f) => (f b -> b) -> (a -> m (f (Either b a))) -> a -> m b
```

Comment:
This new specialized function interleaves does the heavy lifting: it interleaves the monad with an `apo` and a `cata`, all in a single O(n) pass.
The `Either` from apo is used to signal short-circuit (`Left` returns a pre-computed subtree immediately).
It's a shame we can't use `ghylo` for this, and need to write it by hand. This is basically because `Update` can't be `sequence`d.